### PR TITLE
Fix security-context for apiserver - audit logs are supported only in Enterprise version

### PR DESCRIPTION
**Pull request by mihivagyok**
## Description
kind/bug

It fixes the case when the operator installs apiserver OSS version. In case of OSS Calico, it shall run as non-root as audit logs are only supported in Enterprise version.
The change makes apiserver deployment more secure (least privilege).

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by tmjd** [comment]
Have you tested this out in a real system (I'm guessing so because of your other PRs)?
I'm asking because I thought the User and Group IDs had to match the one built into the image but I am by no means confident in that thought. So if you've tried it out and it works then I'm obviously wrong.

File: pkg/render/apiserver.go
Comment (tmjd) [inline pkg/render/apiserver.go:1043]: I think we need the RootContext for all `Variant == Enterprise`, so the `c.cfg.Openshift` should not be part of the if statement.

Comment (mihivagyok) [reply pkg/render/apiserver.go:1043]: Okay, I can make that change. 

So today, the `NewRootContext()` function sets the `RunAsGroup` and `RunAsUser` to `0`, and if it is Openshift, then it sets also `AllowPrivilegeEscalation` and `Privileged` to `true` value.

The Dockerfile defined RunAsGroup and RunAsUser values would be set if we would use the `NewNonRootContext()` function.

My goal here is to run the calico-apiserver as a non-root container and we don't use tigera-enterprise at the moment.

Thank you,
Adam

Comment (hjiawei) [reply pkg/render/apiserver.go:1043]: Am I reading this if condition correct? This condition says that we will use `NewRootContext` only when Calico Enterprise *and* on openshift. It will break on all non-openshift Calico Enterprise cluster. I think we need to remove `&& c.cfg.Openshift` from the if condition.

```
c.cfg.Installation.Variant == operatorv1.TigeraSecureEnterprise &&
		c.cfg.Openshift
```

**Comment by mihivagyok** [comment]
@tmjd I will get back to you with some test result tomorrow! Thanks!

File: pkg/render/apiserver.go
Comment (tmjd) [reply pkg/render/apiserver.go]: Looks good. Thanks for changing that.

**Review by tmjd — APPROVED** [review]
LGTM

**Comment by mihivagyok** [comment]
```
E0216 17:42:12.047708       1 run_server.go:77] open /tmp/ready: permission denied
F0216 17:42:12.047724       1 hooks.go:203] PostStartHook "apiserver-autoregistration" failed: open /tmp/ready: permission denied
```


Seems it is not really working. Try to figure out why:
https://github.com/projectcalico/calico/blob/release-v3.27/apiserver/cmd/apiserver/server/run_server.go#L75

**Comment by mihivagyok** [comment]
@hjiawei Thanks for pointing out!
On the master branch, the Dockerfile is quite different:
https://github.com/projectcalico/calico/blob/master/apiserver/Dockerfile

Also, I don't really understand the comment here. https://github.com/projectcalico/calico/blob/release-v3.27/apiserver/docker-image/Dockerfile.amd64#L8 For me, it seems it does not special with `/code`.

Thanks!

**Comment by hjiawei** [comment]
> Also, I don't really understand the comment here. https://github.com/projectcalico/calico/blob/release-v3.27/apiserver/docker-image/Dockerfile.amd64#L8 For me, it seems it does nothing special with `/code`.

Yeah. You are right about this. It is related to `/tmp` folder permission and not `/code`. We reworked Dockerfiles on the main branch so it is different / unified for multi-arch.

File: pkg/render/apiserver.go
Comment (hjiawei) [inline pkg/render/apiserver.go:1094]: Need `/tmp` folder permission to be 0777 for Calico v3.27 [apiserver Dockerfile](https://github.com/projectcalico/calico/blob/release-v3.27/apiserver/docker-image/Dockerfile.amd64#L11). The `/tmp/ready` flag will write to `/tmp`. I believe this will work with the main Calico branch.

Comment (mihivagyok) [reply pkg/render/apiserver.go:1098]: Hi! I made some fix: I have created an `emptyDir` volume for the `/tmp` so it works now:
```
➜ kubectl get pods -n calico-apiserver
NAME                                READY   STATUS    RESTARTS   AGE
calico-apiserver-58fb6b4bd7-66crp   1/1     Running   0          3m28s
calico-apiserver-58fb6b4bd7-w8ls4   1/1     Running   0          3m28s

➜ kubectl get tigerastatus apiserver
NAME        AVAILABLE   PROGRESSING   DEGRADED   SINCE
apiserver   True        False         False      2m58s
```

Comment (mihivagyok) [reply pkg/render/apiserver.go:1098]: Not sure this is the best, but it can work with the current releases - and hopefully with the future ones. What do you think? Thanks!

**Comment by mihivagyok** [comment]
With latest changes:
- deployment:
```
➜ kubectl get deployment -n calico-apiserver calico-apiserver -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: "2024-02-20T16:35:18Z"
  generation: 1
  labels:
    apiserver: "true"
    app.kubernetes.io/name: calico-apiserver
    k8s-app: calico-apiserver
  name: calico-apiserver
  namespace: calico-apiserver
  ownerReferences:
  - apiVersion: operator.tigera.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: APIServer
    name: default
    uid: c580d54c-30d8-449b-a0bf-5205c90296b2
  resourceVersion: "400266"
  uid: 87600bdc-7bf8-4f50-9c54-d93306caaefc
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      apiserver: "true"
  strategy:
    type: Recreate
  template:
    metadata:
      annotations:
        tigera-operator.hash.operator.tigera.io/calico-apiserver-certs: 5f64c431bd397c8b87c8ebcc7721ac1c9f7206dc
      creationTimestamp: null
      labels:
        apiserver: "true"
        app.kubernetes.io/name: calico-apiserver
        k8s-app: calico-apiserver
      name: calico-apiserver
      namespace: calico-apiserver
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  k8s-app: calico-apiserver
              namespaces:
              - calico-apiserver
              topologyKey: kubernetes.io/hostname
            weight: 100
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  k8s-app: calico-apiserver
              namespaces:
              - calico-apiserver
              topologyKey: topology.kubernetes.io/zone
            weight: 100
      containers:
      - args:
        - --secure-port=5443
        - --tls-private-key-file=/calico-apiserver-certs/tls.key
        - --tls-cert-file=/calico-apiserver-certs/tls.crt
        env:
        - name: DATASTORE_TYPE
          value: kubernetes
        - name: KUBERNETES_SERVICE_HOST
          value: 172.21.0.1
        - name: KUBERNETES_SERVICE_PORT
          value: "443"
        - name: MULTI_INTERFACE_MODE
          value: none
        image: us.icr.io/armada-master/calico/apiserver:v3.27.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /version
            port: 5443
            scheme: HTTPS
          initialDelaySeconds: 90
          periodSeconds: 60
          successThreshold: 1
          timeoutSeconds: 10
        name: calico-apiserver
        readinessProbe:
          exec:
            command:
            - /code/filecheck
          failureThreshold: 5
          initialDelaySeconds: 5
          periodSeconds: 30
          successThreshold: 1
          timeoutSeconds: 10
        resources: {}
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          privileged: false
          runAsGroup: 10001
          runAsNonRoot: true
          runAsUser: 10001
          seccompProfile:
            type: RuntimeDefault
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /calico-apiserver-certs
          name: calico-apiserver-certs
          readOnly: true
        - mountPath: /tmp
          name: cache-volume
      dnsPolicy: ClusterFirst
      nodeSelector:
        kubernetes.io/os: linux
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: calico-apiserver
      serviceAccountName: calico-apiserver
      terminationGracePeriodSeconds: 30
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
      volumes:
      - name: calico-apiserver-certs
        secret:
          defaultMode: 420
          secretName: calico-apiserver-certs
      - emptyDir: {}
        name: cache-volume
status:
  availableReplicas: 2
  conditions:
  - lastTransitionTime: "2024-02-20T16:35:50Z"
    lastUpdateTime: "2024-02-20T16:35:50Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2024-02-20T16:35:19Z"
    lastUpdateTime: "2024-02-20T16:35:50Z"
    message: ReplicaSet "calico-apiserver-58fb6b4bd7" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 2
  replicas: 2
  updatedReplicas: 2
```

- status
```
➜ kubectl get pods -n calico-apiserver
NAME                                READY   STATUS    RESTARTS   AGE
calico-apiserver-58fb6b4bd7-66crp   1/1     Running   0          3m28s
calico-apiserver-58fb6b4bd7-w8ls4   1/1     Running   0          3m28s

➜ kubectl get tigerastatus apiserver
NAME        AVAILABLE   PROGRESSING   DEGRADED   SINCE
apiserver   True        False         False      2m58s
```

File: pkg/render/apiserver.go
Comment (hjiawei) [reply pkg/render/apiserver.go:1098]: Maybe we can keep the securityContext change only for main branch and add the emptyDir part when picking to an operator release branch. I believe we don't need to mount the emptyDir for main Calico apiserver due to the reasons I mentioned earlier. An alternative options is to fix the `/tmp` dir permission in Calico v3.27 branch so we don't need to mount the emptyDir.

Comment (hjiawei) [reply pkg/render/apiserver.go:1098]: Created https://github.com/projectcalico/calico/pull/8570 for Calico OSS v3.27. We can remove the `/tmp` emptyDir mount from operator once this change gets in.

Comment (hjiawei) [reply pkg/render/apiserver.go:1098]: @mihivagyok `calico/apiserver:v3.27.2-19-g7cc1fd061b7f` should have the `/tmp` folder permission fix. Can you give it a try?

Comment (mihivagyok) [reply pkg/render/apiserver.go:1098]: Hi @hjiawei ! Thank you very much! Sure, I will give a try and let you know!

Comment (mihivagyok) [reply pkg/render/apiserver.go:1098]: It works!
```
Version:      v3.27.2-19-g7cc1fd061b7f
Build date:   2024-02-29T19:21:31+0000
Git tag ref:  v3.27.2-19-g7cc1fd061
Git commit:   7cc1fd061
...
I0301 16:51:37.450279       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0301 16:51:37.452995       1 run_server.go:80] apiserver is ready.
```

and stays healthy! Thank you very much!

Then I remove my last commit. Or what do you suggest?
Does semaphore get the calico OSS v.3.27 fix automatically?

Thanks!

Comment (hjiawei) [reply pkg/render/apiserver.go:1098]: Thank you for testing this @mihivagyok! I think you can remove the last emptyDir commit. calico/apiserver change will be in Calico v3.27.3. It doesn't flow back to current release versions automatically.

Once you have this merged into the operator main branch. You also need to pick it back to operator release v1.33 and v1.32 release branches (may need to confirm with operator maintainers).

**Comment by tmjd** [comment]
AFAICT we're still waiting to drop the last/2nd commit that added an emptyDir volume.

**Comment by mihivagyok** [comment]
@tmjd Let me revert those. Sorry for the delay!

File: pkg/render/apiserver.go
Comment (mihivagyok) [reply pkg/render/apiserver.go:1094]: Thank you!

**Comment by mihivagyok** [comment]
All tests are passed :) Thank you!

**Comment by tmjd** [comment]
Thank you @mihivagyok for this PR

**Pull request merged**
